go-owasm/src/lib.rs

```diff
@@ -161,21 +168,22 @@ fn run(code: &[u8], gas_limit: u32, span_size: i64, is_prepare: bool, env: Env)
             }),
             "read_calldata" => func!(|ctx: &mut Ctx, ptr: i64| -> Result<i64, Error> {
                 let vm: &mut vm::VMLogic = unsafe { &mut *(ctx.data as *mut vm::VMLogic) };
-                let span_size = vm.get_span_size() as usize;
-                // TODO: span_size bound check
-                let mut mem: Vec<u8> = Vec::with_capacity(span_size);
-                let mut calldata = Span::create_writable(mem.as_mut_ptr(), span_size);
+                let span_size = vm.get_span_size();
+                let mut mem: Vec<u8> = Vec::with_capacity(span_size as usize);
+                let mut calldata = Span::create_writable(mem.as_mut_ptr(), span_size as usize);
                 vm.get_calldata(&mut calldata)?;
+                require_mem_range( ctx.memory(0).size().bytes().0, (ptr +span_size) as usize )?;
```

Comment:
Let's move this check to right after when `span_size` is created. also make sure code is properly formatted.

ditto for all places below.